File: labs/control-ui/api-shadow/multi_outbound_router.ts
export type OutboundPolicyKind = 'DIRECT' | 'PROXY' | 'REJECT';

export interface OutboundPolicyDecision {
  type: OutboundPolicyKind;
  tag?: string;
}

export interface OutboundRouteRule {
  pattern: string;
  isSuffix: boolean;
  policy: OutboundPolicyDecision;
}

export class MultiOutboundRouter {
  private rules: OutboundRouteRule[] = [];
  private outboundWeights = new Map<string, number>();
  private rrCounter = 0;
  public defaultPolicy: OutboundPolicyDecision;
  constructor(defaultPolicy: OutboundPolicyDecision = { type: 'DIRECT' }) {
    this.defaultPolicy = defaultPolicy;
  }

  addRule(pattern: string, isSuffix: boolean, policy: OutboundPolicyDecision): void {
    this.rules.push({
      pattern: pattern.trim().toLowerCase(),
      isSuffix: Boolean(isSuffix),
      policy,
    });
  }

  registerOutbound(tag: string, weight: number): void {
    this.outboundWeights.set(tag, Math.max(1, weight));
  }

  matchTarget(host: string): OutboundPolicyDecision {
    const clean = host.trim().toLowerCase();
    for (const r of this.rules) {
      if (r.isSuffix) {
        if (clean === r.pattern || clean.endsWith(`.${r.pattern}`)) {
          return r.policy;
        }
      } else if (clean === r.pattern) {
        return r.policy;
      }
    }
    return this.defaultPolicy;
  }

  selectBalancedOutbound(outbounds: string[]): string | undefined {
    if (outbounds.length === 0) return undefined;
    const chosen = outbounds[this.rrCounter % outbounds.length];
    this.rrCounter++;
    return chosen;
  }
}

File: labs/mobile/donor-archive/tunnel/MultiOutboundRouter.kt
package com.luminet.android.tunnel

sealed class AndroidOutboundPolicy {
    object Direct : AndroidOutboundPolicy()
    data class Proxy(val tag: String) : AndroidOutboundPolicy()
    object Reject : AndroidOutboundPolicy()
}

data class AndroidOutboundRule(
    val pattern: String,
    val isSuffix: Boolean,
    val policy: AndroidOutboundPolicy
)

class MultiOutboundRouter(private val defaultPolicy: AndroidOutboundPolicy = AndroidOutboundPolicy.Direct) {
    private val rules = mutableListOf<AndroidOutboundRule>()
    private val outboundWeights = mutableMapOf<String, Int>()
    private var rrCounter = 0

    fun addRule(pattern: String, isSuffix: Boolean, policy: AndroidOutboundPolicy) {
        rules.add(
            AndroidOutboundRule(
                pattern = pattern.trim().lowercase(),
                isSuffix = isSuffix,
                policy = policy
            )
        )
    }

    fun registerOutbound(tag: String, weight: Int) {
        outboundWeights[tag] = if (weight < 1) 1 else weight
    }

    fun matchTarget(host: String): AndroidOutboundPolicy {
        val clean = host.trim().lowercase()
        for (r in rules) {
            if (r.isSuffix) {
                if (clean == r.pattern || clean.endsWith(".${r.pattern}")) {
                    return r.policy
                }
            } else if (clean == r.pattern) {
                return r.policy
            }
        }
        return defaultPolicy
    }

    fun selectBalancedOutbound(outbounds: List<String>): String? {
        if (outbounds.isEmpty()) return null
        val chosen = outbounds[rrCounter % outbounds.size]
        rrCounter++
        return chosen
    }
}
